### PR TITLE
Reduce prover memory usage.

### PR DESCRIPTION
**Pull request by martun**

### CODE REVIEW

File: include/nil/crypto3/zk/snark/arithmetization/plonk/assignment.hpp [hunk 1]
```diff
@@ -147,6 +151,10 @@ namespace nil {
                         return _public_inputs;
                     }
 
+                    public_input_container_type move_public_inputs() const {
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/arithmetization/plonk/assignment.hpp:154]:
```suggestion
                    public_input_container_type move_public_inputs() {

```

File: include/nil/crypto3/zk/snark/arithmetization/plonk/assignment.hpp [hunk 2]
```diff
@@ -164,6 +172,10 @@ namespace nil {
                         return _constants;
                     }
 
+                    constant_container_type move_constants() const {
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/arithmetization/plonk/assignment.hpp:175]:
```suggestion
                    constant_container_type move_constants() {

```

File: include/nil/crypto3/zk/snark/arithmetization/plonk/assignment.hpp [hunk 3]
```diff
@@ -181,6 +193,10 @@ namespace nil {
                         return _selectors;
                     }
 
+                    selector_container_type move_selectors() const {
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/arithmetization/plonk/assignment.hpp:196]:
```suggestion
                    selector_container_type move_selectors() {

```

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 4]
```diff
@@ -66,34 +67,337 @@ namespace nil {
...
+
+                        // Prepare lookup input                        
+                        auto lookup_input_ptr = std::make_unique<std::vector<math::polynomial_dfs<typename FieldType::value_type>>>();
+                        for (const auto &gate : lookup_gates) {
+                            math::expression<DfsVariableType> expr; 
+                            math::polynomial_dfs<typename FieldType::value_type> lookup_selector = plonk_columns.selector(gate.tag_index);
+                            for (const auto &constraint : gate.constraints) {
+                                math::polynomial_dfs<typename FieldType::value_type> l = lookup_selector * (typename FieldType::value_type(constraint.table_id));
+                                theta_acc = theta;
+                                for( std::size_t k = 0; k < constraint.lookup_input.size(); k++){
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:355]:
```suggestion
                                for(std::size_t k = 0; k < constraint.lookup_input.size(); k++){
```

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/prover.hpp [hunk 5]
```diff
@@ -99,37 +99,35 @@ namespace nil {
...
-                            , preprocessed_private_data(preprocessed_private_data)
                             , table_description(table_description)
                             , constraint_system(constraint_system)
-                            , assignments(assignments)
                             , _commitment_scheme(commitment_scheme)
-                            , _polynomial_table(preprocessed_private_data.private_polynomial_table,
-                                                preprocessed_public_data.public_polynomial_table)
+                            , _polynomial_table(new plonk_polynomial_dfs_table<FieldType, typename ParamsType::arithmetization_params>(
+                                std::move(preprocessed_private_data.private_polynomial_table),
+                                preprocessed_public_data.public_polynomial_table))
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/systems/plonk/placeholder/prover.hpp:129]:
```suggestion
                            , _polynomial_table(std::make_unique<plonk_polynomial_dfs_table<FieldType, typename ParamsType::arithmetization_params>>(
                                std::move(preprocessed_private_data.private_polynomial_table),
                                preprocessed_public_data.public_polynomial_table))

```

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 6]
```diff
@@ -66,34 +67,337 @@ namespace nil {
...
+                        math::polynomial_dfs<typename FieldType::value_type> V_L(
+                            basic_domain->m-1,basic_domain->m, FieldType::value_type::zero());
+                        V_L[0] = FieldType::value_type::one();
+                        auto one = FieldType::value_type::one();
+
+                        for (std::size_t k = 1; k <= preprocessed_data.common_data.usable_rows_amount; k++) {
+                            V_L[k] = V_L[k-1];
+                        
+                            typename FieldType::value_type g_tmp = (one + beta).pow(reduced_input.size());
+                            for( std::size_t i = 0; i < reduced_input.size(); i++){
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:267]:
```suggestion
                            for (std::size_t i = 0; i < reduced_input.size(); i++) {
```

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 7]
```diff
@@ -66,34 +67,337 @@ namespace nil {
...
+                                g_tmp *= gamma + reduced_input[i][k-1];
+                            }
+
+                            auto part1 = (one + beta) * gamma;
+                            for( std::size_t i = 0; i < reduced_value.size(); i++){
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:272]:
```suggestion
                            for (std::size_t i = 0; i < reduced_value.size(); i++) {
```

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 8]
```diff
@@ -66,34 +67,337 @@ namespace nil {
...
+                                g_tmp *= part1 + reduced_value[i][k-1] + beta * reduced_value[i][k];
+                            }
+
+                            V_L[k] *= g_tmp;
+
+                            typename FieldType::value_type h_tmp = FieldType::value_type::one();
+                            for (std::size_t i = 0; i < sorted.size(); i++) {
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:279]:
```suggestion
                            for (std::size_t i = 0; i < sorted.size(); i++) {
```

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 9]
```diff
@@ -66,34 +67,337 @@ namespace nil {
...
+
+                    math::polynomial_dfs<typename FieldType::value_type> compute_h(
+                            const std::vector<math::polynomial_dfs<typename FieldType::value_type>>& sorted,
+                            const typename FieldType::value_type& beta,
+                            const typename FieldType::value_type& gamma
+                        ) {
+                        auto one = FieldType::value_type::one();
+
+                        std::vector<math::polynomial_dfs<typename FieldType::value_type>> h_multipliers;
+                        for( std::size_t i = 0; i < sorted.size(); i++){
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:244]:
```suggestion
                        for (std::size_t i = 0; i < sorted.size(); i++) {
```

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 10]
```diff
@@ -66,34 +67,337 @@ namespace nil {
...
+                        return V_L;
+                    }
+
+                    std::unique_ptr<std::vector<math::polynomial_dfs<typename FieldType::value_type>>> prepare_lookup_value(
+                            const math::polynomial_dfs<typename FieldType::value_type>& mask_assignment) {
+                        typename FieldType::value_type theta_acc;
+
+                        // Prepare lookup value
+                        auto lookup_value_ptr = std::make_unique<std::vector<math::polynomial_dfs<typename FieldType::value_type>>>();
+                        for(std::size_t t_id = 0; t_id < lookup_tables.size(); t_id++) {
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:293]:
```suggestion
                        for (std::size_t t_id = 0; t_id < lookup_tables.size(); t_id++) {
```

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 11]
```diff
@@ -66,34 +67,337 @@ namespace nil {
...
+                            const plonk_lookup_table<FieldType> &l_table = lookup_tables[t_id];
+                            const math::polynomial_dfs<typename FieldType::value_type> &lookup_tag = plonk_columns.selector(l_table.tag_index);
+                            for( std::size_t o_id = 0; o_id < l_table.lookup_options.size(); o_id++ ){
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:296]:
```suggestion
                            for (std::size_t o_id = 0; o_id < l_table.lookup_options.size(); o_id++) {
```

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 12]
```diff
@@ -66,34 +67,337 @@ namespace nil {
...
+                                math::polynomial_dfs<typename FieldType::value_type> v = (typename FieldType::value_type(t_id + 1)) * lookup_tag;
+                                theta_acc = theta;
+                                for(std::size_t i = 0; i < l_table.columns_number; i++) {
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:299]:
```suggestion
                                for (std::size_t i = 0; i < l_table.columns_number; i++) {
```

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 13]
```diff
@@ -66,34 +67,337 @@ namespace nil {
...
+                        std::vector<math::polynomial_dfs<typename FieldType::value_type>> g_multipliers;
+                        for( std::size_t i = 0; i < lookup_input.size(); i++) {
+                            g_multipliers.push_back(gamma + lookup_input[i]);
+                        }
+
+                        // We don't use lookup_input after this line.
+                        lookup_input_ptr.reset(nullptr);
+
+                        auto part1 = (one+beta) * gamma;
+                        for( std::size_t i = 0; i < lookup_value.size(); i++ ){
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:224]:
```suggestion
                        for (std::size_t i = 0; i < lookup_value.size(); i++) {
```

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 14]
```diff
@@ -66,34 +67,337 @@ namespace nil {
...
+
+                        auto& lookup_value = *lookup_value_ptr;
+                        auto& lookup_input = *lookup_input_ptr;
+ 
+                        auto g = math::polynomial_dfs<typename FieldType::value_type>::one();
+                        auto one = FieldType::value_type::one();
+                        g *= (one + beta).pow(lookup_input.size());
+
+                        std::vector<math::polynomial_dfs<typename FieldType::value_type>> g_multipliers;
+                        for( std::size_t i = 0; i < lookup_input.size(); i++) {
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:216]:
```suggestion
                        for (std::size_t i = 0; i < lookup_input.size(); i++) {
```

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 15]
```diff
@@ -66,34 +67,337 @@ namespace nil {
...
+                        g *= V_L;
+                        h *= V_L_shifted;
+                        g -= h;
+                        h = math::polynomial_dfs<typename FieldType::value_type>(); // just clean the memory of h.
+                        g *= (preprocessed_data.q_last + preprocessed_data.q_blind) - one_polynomial;
+                        F_dfs[2] = std::move(g);
+
+                        F_dfs[3] = zero_polynomial;
+
+                        for( std::size_t i = 1; i < sorted.size(); i++){
```
Comment (x-mass) [inline include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:183]:
```suggestion
                        for (std::size_t i = 1; i < sorted.size(); i++) {
```

**Review by ETatuzova — APPROVED** [review]

File: include/nil/crypto3/zk/snark/systems/plonk/placeholder/prover.hpp (hunk 5 above)
Comment (martun) [reply include/nil/crypto3/zk/snark/systems/plonk/placeholder/prover.hpp:129]: This one doesn't make a difference.

**Pull request merged**